3.4 support for 3.3

sherfert opened the pull request.

sherfert committed:
Starting to support 3.3 in 3.4

Creating Compatibility class that uses 3.3 until we get a
LogicalPlanState and then switches to 3.4. There are still many TODOs
left in place.

sherfert committed:
CostCompatibility for 3.3

sherfert committed:
CompilerEngineDelegator and CompitibilityCache

sherfert committed:
Starting to convert LogicalPlan(states) and getting frustrated.

sherfert committed:
Removing 3.2 from CypherComparisonSupport

Also refactor-renaming Cost3.3 to Cost3.4 (because that is what it
actually was).

sherfert committed:
Converting 3.3 expressions to 3.4 expressions

using a bottomUpRewriterWithArgs and reflection.

sherfert committed:
More tests and fixes for expression converting

sherfert committed:
Converting Logical Plans

sherfert committed:
Memoization of Constructors.

sherfert committed:
Tests with Reflections library. Coding many special conversion cases.

sherfert committed:
Converting all LogicalPlans with all their special snowflakes

sherfert committed:
Fixed Foldable getting children

Our bottomUpWithArgs Rewriter requires a different match order
for children.

sherfert committed:
SemanticTableConverter that takes Expression 3.3->3.4 mapping

sherfert committed:
Infrastructure for filling a Map of 3.3->3.4 expressions

sherfert committed:
Filling a Map of 3.3->3.4 expressions; converting SemanticTable

sherfert committed:
PlannerQueryWrapper & Filtering SemanticTable during conversion

The PlannerQueryWrapper provides a minimal implementation of the
parts of a PlannerQuery that are actually used inside the runtime.

While converting the semantic table from 3.3 to 3.4, we keep only
those entries that are part of the LogicalPlan.

sherfert committed:
Converting returnColumns of Statement.

sherfert committed:
Removing unused def from BaseState

sherfert committed:
Removing TODOs

sherfert committed:
Revert "Removing unused def from BaseState"

This reverts commit e361af9c3fa4613bcaedc4d1a39fc878b67894bb.

sherfert committed:
Removing unused typeInfo from Argument (logical plan)

sherfert committed:
Logging logical and physical planning and notifications when running 3.3

sherfert committed:
Many minor conversion bugfixes.

sherfert committed:
Including 3.3 in Acceptance tests. Fixing TestConfigurations.

sherfert committed:
More bug and test fixes

sherfert committed:
Bugfix where plans don't have an ID assigned.

sherfert committed:
Merge branch '3.4' into 3.4-support-3.3

sherfert committed:
Removed forgotten TODO

sherfert committed:
Update compiler dependency to 3.3.1

sherfert committed:
Merge branch '3.4-update-3.3' into 3.4-support-3.3

sherfert committed:
Fix compiler errors after update to 3.3.1

sherfert committed:
Fixing Acceptance tests after update to 3.3.1

sherfert committed:
Running compatibility tests for 3.3

sherfert committed:
Making Checkstyle happy

sherfert committed:
Merge remote-tracking branch 'origin/3.4' into 3.4-support-3.3

systay commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/CompatibilityCache.scala:
```diff
@@ -67,20 +70,31 @@ class CommunityCompatibilityFactory(graph: GraphDatabaseQueryService, kernelMoni
+  override def create(spec: PlannerSpec_v3_3, config: CypherCompilerConfiguration) =
+    (spec.planner, spec.runtime) match {
+      case (CypherPlanner.rule, _) =>
+        throw new InvalidArgumentException("The rule planner is no longer a valid planner option in Neo4j 3.3. If you need to use it, please compatibility mode Cypher 3.1")
```
Typo in the message "If you need to use it, please `select` compatibility mode Cypher 3.1"

systay commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/Compatibility.scala:
```diff
@@ -0,0 +1,282 @@
+import scala.collection.JavaConverters._
+import scala.util.Try
+
+trait Compatibility[CONTEXT3_3 <: v3_3.phases.CompilerContext,
```
There seems to be a lot of code duplication between this file and https://github.com/neo4j/neo4j/blob/3.4/community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/Compatibility.scala

Just looking at it briefly it's not clear that is necessary duplication. I would really like it if we can minimize it.

systay commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/CostCompatibility.scala:
```diff
@@ -0,0 +1,93 @@
+import org.neo4j.kernel.monitoring.{Monitors => KernelMonitors}
+import org.neo4j.logging.Log
+
+case class CostCompatibility[CONTEXT3_3 <: v3_3.phases.CompilerContext,
```
This is the only implementation of the `Compatibility[CONTEXT3_3, CONTEXT3_4, T]` trait. Why do we have this division? The trait is not only declaring an interface, it contains a whole lot of code. I believe this made sense back in the days where we had both CostCompatablity and RuleCompatability, but I don't think we should perpetuate that pattern any longer.

systay commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/LogicalPlanConverter.scala:
```diff
@@ -0,0 +1,224 @@
+import scala.collection.mutable.{HashMap => MutableHashMap}
+import scala.util.{Failure, Success, Try}
+
+object LogicalPlanConverter {
```
In general, the `.instanceOf[]` calls make the big match clause difficult to read. I would like to see the children cast into variables that we can then use to create the new plan. Example:
```
        case (plan: plansV3_3.TriadicSelection, children: Seq[AnyRef]) =>
          plansV3_4.TriadicSelection(children(1).asInstanceOf[LogicalPlanV3_4],
            children(5).asInstanceOf[LogicalPlanV3_4],
            children(0).asInstanceOf[Boolean],
            children(2).asInstanceOf[IdNameV3_4],
            children(3).asInstanceOf[IdNameV3_4],
            children(4).asInstanceOf[IdNameV3_4])(new PlannerQueryWrapper(plan.solved))
``` 

could instead be:

```
        case (plan: plansV3_3.TriadicSelection, children: Seq[AnyRef]) =>
          val lhs = children(1).asInstanceOf[LogicalPlanV3_4]
          val rhs = children(5).asInstanceOf[LogicalPlanV3_4]
          val positivePredicate = children(0).asInstanceOf[Boolean]
          val sourceId = children(2).asInstanceOf[IdNameV3_4]
          val seenId = children(3).asInstanceOf[IdNameV3_4]
          val targetId = children(4).asInstanceOf[IdNameV3_4]

          plansV3_4.TriadicSelection(lhs, rhs, positivePredicate, sourceId, seenId, targetId)(new PlannerQueryWrapper(plan.solved))
```

sherfert replied:
Would naming the arguments work for you too? Like

```
plansV3_4.TriadicSelection(lhs = children(1).asInstanceOf[LogicalPlanV3_4],
            rhs = children(5).asInstanceOf[LogicalPlanV3_4],
            positivePredicate = children(0).asInstanceOf[Boolean],
            sourceId = children(2).asInstanceOf[IdNameV3_4],
            seenId = children(3).asInstanceOf[IdNameV3_4],
            targetId = children(4).asInstanceOf[IdNameV3_4])
```

systay commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/PipeExecutionPlanBuilder.scala:
```diff
@@ -76,7 +76,7 @@ class PipeExecutionPlanBuilder(clock: Clock,
       monitors = monitors,
       recurse = p => buildPipe(p),
-      readOnly = plan.solved.all(_.queryGraph.readOnly),
+      readOnly = plan.solved.readOnly,
```
Niice :)

systay commented on community/cypher/util-3.4/src/main/scala/org/neo4j/cypher/internal/util/v3_4/Rewritable.scala:
```diff
@@ -246,3 +261,47 @@ object bottomUp {
+
+    @tailrec
+    private def rec(
+                     stack: mutable.ArrayStack[(List[AnyRef], mutable.MutableList[AnyRef])]
```
Why so many newlines inside the parens?

systay reviewed (changes requested):
Really good start. A few improvement suggestions

sherfert committed:
Minor review comments

sherfert committed:
Merged CostCompatibility and Compatibility for both 3.3 and 3.4

3.3 and 3.4 still have their separate Compatibility case class.

sherfert committed:
Abstracting over 3.3 and 3.4 Compatibility, step 1

sherfert committed:
Minor code cleanups. Licence header.

sherfert committed:
More minor code cleanups.

craigtaverner commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/LogicalPlanConverter.scala:
```diff
@@ -0,0 +1,225 @@
+    new RewritableAny[astV3_3.ASTNode](ast).rewrite(new LogicalPlanRewriter(expressionMap, isImportant), Seq.empty).asInstanceOf[T]
+  }
+
+  private val constructors = new ThreadLocal[MutableHashMap[(String, String, String), Constructor[_]]]() {
```
Why ThreadLocal? Does that not mean every thread (every query) will have to re-create the cache? Would it not be better to have a cache that survives the duration of one query? Perhaps a concurrent mutable hashmap?

sherfert replied:
I hope threads are actually pooled, but not sure. Anyway, it is a tradeoff. Maybe something worth measuring?

craigtaverner commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/LogicalPlanConverter.scala:
```diff
@@ -0,0 +1,225 @@
+
+        case (spp: irV3_3.ShortestPathPattern, children: Seq[AnyRef]) =>
+          val sp3_4 = convertASTNode[expressionsV3_4.ShortestPaths](spp.expr, expressionMap, isImportant)
+          irV3_4.ShortestPathPattern(children(0).asInstanceOf[Option[IdNameV3_4]], children(1).asInstanceOf[irV3_4.PatternRelationship], children(2).asInstanceOf[Boolean])(sp3_4)
```
Could this also have used convertVersion? Since parameter order is unchanged?

sherfert replied:
It doesn't work because of the second set of arguments `(sp3_4)`

craigtaverner commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/LogicalPlanConverter.scala:
```diff
@@ -0,0 +1,225 @@
+
+    private val rewriter: RewriterWithArgs = bottomUpWithArgs { before =>
+      val rewritten = RewriterWithArgs.lift {
+        case (plan: plansV3_3.Argument, children: Seq[AnyRef]) =>
```
Would it not make more sense to name `children` into `parameters`.

sherfert replied:
I think it both makes sense. From a constructor perspective we are talking about parameters, from a tree perspective we are talking about children.

craigtaverner commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/helpers.scala:
```diff
@@ -0,0 +1,170 @@
+    val startPosition = logicalPlan.startPosition.map(as3_4 _)
+    val plannerName = as3_4(logicalPlan.plannerName)
+
+    def isImportant(expression: ExpressionV3_3) : Boolean = logicalPlan.maybeSemanticTable.map(_.seen(expression)).getOrElse(false)
```
The `map(_.seen(expression)).getOrElse(false)` could be replaced with `exists(_.seen(expression))`

craigtaverner commented on community/cypher/cypher-logical-plans-3.4/src/main/scala/org/neo4j/cypher/internal/v3_4/logical/plans/ProcedureSignature.scala:
```diff
@@ -70,7 +72,20 @@ case class FieldSignature(name: String, typ: CypherType, default: Option[CypherV
+      false
+    } else {
+      val other = obj.asInstanceOf[ProcedureAccessMode]
+      util.Arrays.equals(this.allowed.asInstanceOf[Array[Object]], other.allowed.asInstanceOf[Array[Object]])
```
Should this perhaps do a set equals? Order should not matter, right?

sherfert replied:
I think you should be right. Although there is the usual lack of documentation, so I am unsure how `allowed` is actually used. @OliviaYtterbrink ?

craigtaverner commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/LatestRuntimeVariablePlannerCompatibility.scala:
```diff
@@ -0,0 +1,194 @@
+      }.asJava)
+    }
+  }
+
```
Two consecutive lines?

craigtaverner commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/LatestRuntimeVariablePlannerCompatibility.scala:
```diff
@@ -0,0 +1,194 @@
+
+  protected def logStalePlanRemovalMonitor(log: InfoLogger) = new AstCacheMonitor[STATEMENT] {
+    override def cacheDiscard(key: STATEMENT, userKey: String, secondsSinceReplan: Int) {
+      log.info(s"Discarded stale query from the query cache: $userKey")
```
The code this came from also logged the secondsSinceReplan, so this should be put back:

```
log.info(s"Discarded stale query from the query cache after ${secondsSinceReplan} seconds: $userKey")
```

craigtaverner commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/Compatibility.scala:
```diff
@@ -0,0 +1,205 @@
+    val inputPositionV3_3 = helpers.as3_3(preParsedQuery.offset)
+    val inputPositionV3_4 = preParsedQuery.offset
+    val notificationLoggerV3_3 = new RecordingNotificationLoggerV3_3(Some(inputPositionV3_3))
+    val notificationLoggerV3_4 = new RecordingNotificationLogger(Some(inputPositionV3_4))
```
Would look slightly nicer to suffix V3_4 onto the classname as done elsewhere..

craigtaverner commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/Compatibility.scala:
```diff
@@ -0,0 +1,205 @@
+        val preparedQuery = compiler.normalizeQuery(syntacticQuery, contextV3_3)
+        val cache = provideCache(cacheAccessor, cacheMonitor, planContextV3_3, planCacheFactory)
+        val statisticsV3_4 = GraphStatisticsWrapper(planContextV3_3.statistics)
+        val isStale = (plan: ExecutionPlan_v3_4) => plan.isStale(planContextV3_3.txIdProvider, statisticsV3_4)
```
Might read better to use `planContextV3_4.txIdProvider`

craigtaverner commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/GraphStatisticsWrapper.scala:
```diff
@@ -0,0 +1,87 @@
+  }
+
+
+  override def nodesAllCardinality() = ??? // TODO: Implement once we rely on 3.3.1 that has this method
```
We already do depend on 3.3.1, perhaps this should be implemented?

craigtaverner commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/Compatibility.scala:
```diff
@@ -0,0 +1,205 @@
+        //Prepare query for caching
+        val preparedQuery = compiler.normalizeQuery(syntacticQuery, contextV3_3)
+        val cache = provideCache(cacheAccessor, cacheMonitor, planContextV3_3, planCacheFactory)
+        val statisticsV3_4 = GraphStatisticsWrapper(planContextV3_3.statistics)
```
Do we need the GraphStatisticsWrapper at all? The v3_3 statistics are themselves converted back into v3_4 calls using the v3_3/TransactionBoundGraphStatistics calls. This looks like double wrapping v3_4(v3_3(v3_4)). Surely we can just pass v3_4 in directly?

craigtaverner commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/SemanticTableConverter.scala:
```diff
@@ -0,0 +1,99 @@
+      new mutable.HashMap[String, utilV3_4.PropertyKeyId],
+      convert(table.resolvedRelTypeNames))
+  }
+
```
multiple newlines

craigtaverner commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/SemanticTableConverter.scala:
```diff
@@ -0,0 +1,99 @@
+    }
+    res
+  }
+
```
Multiple newlines

craigtaverner commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/SemanticTableConverter.scala:
```diff
@@ -0,0 +1,99 @@
+
+  private def convert(typeInfoV3_3: frontendV3_3.ExpressionTypeInfo): frontendV3_4.semantics.ExpressionTypeInfo = typeInfoV3_3 match {
+    case frontendV3_3.ExpressionTypeInfo(specified, expected) =>
+      frontendV3_4.semantics.ExpressionTypeInfo(convert(specified), expected.map(convert _))
```
Could delete the ` _`

craigtaverner commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/SemanticTableConverter.scala:
```diff
@@ -0,0 +1,99 @@
+  }
+
+  private def convert(specified: frontendV3_3.symbols.TypeSpec): utilV3_4.symbols.TypeSpec = {
+      new utilV3_4.symbols.TypeSpec(specified.ranges.map(convert _))
```
Could delete the ` _`

craigtaverner commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/SemanticTableConverter.scala:
```diff
@@ -0,0 +1,99 @@
+
+  private def convert(range: frontendV3_3.symbols.TypeRange): utilV3_4.symbols.TypeRange = range match {
+    case frontendV3_3.symbols.TypeRange(lower, upper) =>
+      utilV3_4.symbols.TypeRange(convert(lower), upper.map(convert _))
```
Could delete the ` _`

craigtaverner commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v3_3/QueryTransactionalContext.scala:
```diff
@@ -0,0 +1,44 @@
+import org.neo4j.cypher.internal.compiler.v3_3.spi._
+import org.neo4j.kernel.impl.factory.DatabaseInfo
+
+trait QueryTransactionalContext {
```
This entire file should not be needed.

craigtaverner commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v3_3/TransactionBoundPlanContext.scala:
```diff
@@ -0,0 +1,196 @@
+
+import scala.collection.JavaConverters._
+
+class TransactionBoundPlanContext(tc: TransactionalContextWrapper, logger: InternalNotificationLogger)
```
To get rid of TransactionalContextWrapper, we should pass in instead a `readOperationsSupplier` of type `() => ReadOperations`

craigtaverner commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v3_3/TransactionBoundTokenContext.scala:
```diff
@@ -0,0 +1,72 @@
+import org.neo4j.kernel.api.exceptions.{LabelNotFoundKernelException, PropertyKeyNotFoundException, RelationshipTypeNotFoundException}
+import org.neo4j.kernel.impl.api.operations.KeyReadOperations
+
+abstract class TransactionBoundTokenContext(statement: => Statement) extends TokenContext {
```
Rather pass in instead a readOperationsSupplier of type () => ReadOperations

craigtaverner commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v3_3/TransactionalContextWrapper.scala:
```diff
@@ -0,0 +1,75 @@
+import org.neo4j.kernel.impl.factory.DatabaseInfo
+import org.neo4j.kernel.impl.query.TransactionalContext
+
+case class TransactionalContextWrapper(tc: TransactionalContext) extends QueryTransactionalContext {
```
This class can be deleted

craigtaverner commented on enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/CompositeNodeKeyConstraintAcceptanceTest.scala:
```diff
@@ -282,63 +269,52 @@ class CompositeNodeKeyConstraintAcceptanceTest extends ExecutionEngineFunSuite w
     }
   }
 
+
```
Why?

craigtaverner committed:
Fixed failing tests, and removed unnecessary code

fickludd commented on community/cypher/cypher/src/test/scala/org/neo4j/cypher/internal/compatibility/v3_4/LFUCacheTest.scala:
```diff
@@ -19,6 +19,7 @@
+import org.neo4j.cypher.internal.compatibility.LFUCache
 import org.neo4j.cypher.internal.util.v3_4.test_helpers.CypherFunSuite
 
 class LFUCacheTest extends CypherFunSuite {
```
This should be move along with `LFUCache`

fickludd commented on community/cypher/cypher-logical-plans-3.4/src/main/scala/org/neo4j/cypher/internal/v3_4/logical/plans/Argument.scala:
```diff
@@ -26,18 +26,17 @@ import org.neo4j.cypher.internal.util.v3_4.symbols._
   * Produce a single row with the contents of argument
   */
 case class Argument(argumentIds: Set[IdName] = Set.empty)(val solved: PlannerQuery with CardinalityEstimation)
-                   (val typeInfo: Map[String, CypherType] = argumentIds.map( id => id.name -> CTNode).toMap)
```
Yes!

sherfert committed:
Review comments

fickludd commented on community/cypher/cypher-logical-plans-3.4/src/main/scala/org/neo4j/cypher/internal/v3_4/logical/plans/LogicalPlan.scala:
```diff
@@ -71,6 +71,9 @@ abstract class LogicalPlan
   override def rememberMe(old: AnyRef): Unit = _id = old.asInstanceOf[LogicalPlan]._id
 
+  // This should only be used for converting old versions of LogicalPlans
+  def setIdTo(id: LogicalPlanId) : Unit = _id = Some(id)
```
The logical plan abstraction is leaky... we might want to revisit that later.

sherfert replied:
Yeah, it's kinda sucky, I agree

fickludd commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/SemanticTableConverter.scala:
```diff
@@ -0,0 +1,99 @@
+    case frontendV3_3.symbols.CTPoint => utilV3_4.symbols.CTPoint
+    case frontendV3_3.symbols.CTRelationship => utilV3_4.symbols.CTRelationship
+    case frontendV3_3.symbols.CTString => utilV3_4.symbols.CTString
+  }
```
This could possibly be share implementation with the type conversion in the `LogicalPlanConverter`, if that's reasonably smooth.

sherfert replied:
We decided against that, given that here the number of cases is reasonably small, so that we do not want to pay the reflection overhead.

fickludd replied:
I was thinking to call the other way, but no problem - do as you want =)

fickludd reviewed (commented):
Awesome work in getting this functionality implemented. I think working in this area exposes some wrinkly parts of the codebase, that you should not have to fix in this PR. Let's get back here later and unwrinkle some more.

sherfert committed:
Convert trait to abstract class: LatestRuntimeVariablePlannerCompat...

By doing that, all constructor arguments are passed explicitly to
the super constructor and we can be sure those are initialized at
construction time. All remaining abstract vals are changed to be used
only in defs, not in concrete vals anymore. This makes this class not
be a fragile base class anymore.

sherfert committed:
Merge remote-tracking branch 'origin/3.4' into 3.4-support-3.3

sherfert committed:
Fix compilation error. Move LFUCacheTest to correct package.

sherfert committed:
Fix test failures around query caching.

sherfert committed:
Merge remote-tracking branch 'origin/3.4' into 3.4-support-3.3

sherfert committed:
Fix compilation error

sherfert committed:
using set instead of array equality

craigtaverner committed:
Make sporadic failure determinisitic

craigtaverner committed:
Get planning of 3.3 to save 3.4 compatible graph statistics

craigtaverner committed:
Missing license header
